### ISSUE TITLE
Expose `/etc/nsswitch.conf` and `/etc/protocols` etc to sandboxed nix builds?

### ISSUE THREAD

**Issue by bennofs** [NixOS/nix#1238]
Right now, the following derivation fails in the nix sandbox:

```
with (import <nixpkgs> { config = {}; });

stdenv.mkDerivation {
  name = "networking";
  buildCommand = ''
    ${strace}/bin/strace getent protocols tcp > $out
  '';
}
```

This means that even packages that run their own server on localhost for testing fail, because `getprotobyname` fails.

**Comment by bennofs** [comment]
See alephcloud/hs-configuration-tools#40 for a package affected by this

**Comment by copumpkin** [comment]
Seems reasonable to me. Would need to switch on host system since these don't exist on Darwin.

My inclination would be to have them contain super boring predefined values that Nix defines somewhere, rather than passing through the host. That improves determinism, and we likely don't need anything fancy someone might have added to `nsswitch.conf` (which even supports loadable libraries!) inside a test that wants to hit localhost.

**Comment by domenkozar** [comment]
+1, just being able to do reverse dns on `/etc/hosts` would be a good start for some tests.

**Comment by domenkozar** [comment]
If I understand correctly this needs `hosts: files` entry for `/etc/nsswitch.conf` besides https://github.com/NixOS/nix/blob/master/src/libstore/build.cc#L1990

**Comment by bennofs** [comment]
@domenkozar yes, and we should probably also use `files` for the other databases, like `protocols` and `services`.

**Comment by bennofs** [comment]
(And add `/etc/protocols` + `/etc/services`)

**Comment by edolstra** [comment]
I think this can already be done using `--sandbox-paths /etc/protocols:...`.

**Comment by copumpkin** [comment]
What I was proposing was for them to be dummy files though, much like how `/etc/passwd` doesn't actually contain real `passwd` entries but is populated in `build.cc`.

**Comment by edolstra** [comment]
Yes, but especially for /etc/services it's not clear what the dummy contents should be.

**Comment by bennofs** [comment]
@edolstra is there no standard `/etc/services`? Just use the IANA standard?

**Comment by jonringer** [comment]
Just ran into another use case where this would have been incredibly helpful to have an `/etc/protocols` for the pythonPackages.eventlet package, which references it for the `socket.getprotobyname('tcp')` python command. (similar to @Mic92's example)

**Comment by nobodyinperson** [comment]
Another use case is to get git annex working inside the sandbox of a fixed-output-derivation: https://gitlab.com/nobodyinperson/nixconfig/-/snippets/4893095
